### PR TITLE
AP-3509 Gracefully prevent calling GET on the same assessment twice

### PR DESCRIPTION
**Pull request by stephenrichards**
## Gracefully prevent calling GET on the same assessment twice

Users reported getting exceptions when calling GET more than once for the same assessment.

This was because during the assessment process, eligibilty records are created and were failing because they already existed.  In order to enable calling GET on the assessment to be truly idempotent, we wold have had to delete any eligibilty records, remarks records and zeroise all the totals on the summary records.

For that reason it was felt better to gracefully refuse to carry out a seconds assesment as this would not be a normal use-case anyay, and return and explanatory message.

The swagger documentation has been updated to reflect this.

The error being raised with this was:
```
PG::UniqueViolation: ERROR: duplicate key value violates unique constraint "eligibilities_unique_type_ptc" 
...
```

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/controllers/assessments_controller.rb [hunk 1]
```diff
@@ -8,8 +8,11 @@ def create
   end
 
   def show
-    if assessment_incomplete?
+    case analyse_assessment_status
+    when :incomplete
       render json: Decorators::ErrorDecorator.new(incomplete_message).as_json, status: :unprocessable_entity
+    when :not_pending
```
Comment (jsugarman) [inline app/controllers/assessments_controller.rb:14]: would `:already_assessed` or similar be clearer?

File: spec/requests/swagger_docs/v5/assessments_spec.rb [hunk 2]
```diff
@@ -67,6 +67,11 @@
       tags "Assessment"
       produces "application/json"
 
+      description <<~DESCIPTION.chomp
+        This endpoint will cause an assessment to be calculated using all the data that has been uploaded to the
+        other endpoints and return the result.  You call this endpoint twice for the same assessment_id.
```
Comment (jsugarman) [inline spec/requests/swagger_docs/v5/assessments_spec.rb:72]: should this be 
```suggestion
        other endpoints and return the result.  You cannot call this endpoint twice for the same assessment_id.
```

File: swagger/v5/swagger.yaml [hunk 3]
```diff
@@ -104,6 +104,9 @@ paths:
       summary: show assessment
       tags:
       - Assessment
+      description: |-
+        This endpoint will cause an assessment to be calculated using all the data that has been uploaded to the
+        other endpoints and return the result.  You call this endpoint twice for the same assessment_id.
```
Comment (jsugarman) [inline swagger/v5/swagger.yaml:109]: [ditto](https://github.com/ministryofjustice/check-financial-eligibility/pull/1120/files#r981402693)

**Review by jsugarman — APPROVED** [review]
LGTM 👍

**Comment by colinbruce** [comment]
Is it worth having a chat with the EFE team about their requirements?   I believe they were originally looking at a DB-less system that may need to render our info twice and, therefore, GET the assessment twice.

Rather than provide a nice error, would simply return the assessment record if it exists, or create it first if it doesn't make more sense?

**Comment by jsugarman** [comment]
Sure, I think a GET should be repeatable but I don't understand what the implications are for say retrieving the assessment a few days later. is there some date sensitivity?! In addition if I GET the assessment, then add some more outgoings or other data, does the GET need to do some kind of update as Stephen alludes to...

> In order to enable calling GET on the assessment to be truly idempotent, we wold have had to delete any eligibilty records, remarks records and zeroise all the totals on the summary records.

Or do we need to prevent adding data on other endpoints once the GET has been performed for a given assessment?!

**Comment by cpjmcquillan** [comment]
_disclaimer: I haven't look at any of the code, and I'm going entirely off what is being described as the system's behaviour_

If a GET request is doing anything other than retrieving some data, should it be a GET request at all?

It sounds like we're in a bit of a pickle and the API conflates _doing_ an assessment with _retrieving_ an assessment result.

**Comment by colinbruce** [comment]
> I don't want to block this change if the error is impacting or blocking another team, but I do think this is something that is important and we should resolve —ideally before the API is used more widely.

I think taking our time is okay - EFE raised a query in Slack about why they couldn't call the endpoint twice and I added a backlog ticket to discuss and identify a solution.

Stephen picked it up and proposed this... I think we should probably take our time and discuss as a team with EFE to take a more measured approach to a solution

**Comment by colinbruce** [comment]
Marked as OnHold pending an Apply/EFE Lead Dev meeting to discuss ways of working/best way of implementing this

**Comment by colinbruce** [comment]
Closing as partner/civil have come up with a one-shot api that negates the need for this

**Pull request closed without merging**